_blog/cpp-reflection.md

```diff
@@ -0,0 +1,529 @@
+---
+layout: blog-detail
+name: "Reflection in C++. We waited for it! Or did we?"
+image: "/assets/blog/reflection.png"
+description: "An overview of reflection in C++26 with examples"
+tags:
+  - "C++"
+---
+# Reflection in C++. We waited for it! Or did we?
+## What is reflection?
+Before start talking about reflection in my beloved C++, we need to figure out what is reflection.
+
+> According to [Wikipedia](https://en.wikipedia.org/wiki/Reflective_programming):  
+> In computer science, **reflective programming** or **reflection** is the ability of a process to examine, introspect, and modify its own structure and behavior.
+
+In simple words, reflection allows the program to ask itself: "what do I consist of."  
+Talking a bit more practical reflection allows us to figure out what fields the object has, which functions and maybe modify itself:
+```kotlin
+package dev.spgc
+  
+import kotlin.reflect.KProperty1  
+import kotlin.reflect.full.companionObject  
+import kotlin.reflect.full.declaredMemberFunctions  
+  
+class Car(  
+    val amountOfWheels: Int,  
+    var ownerName: String,  
+    var amountOfCrashes: Int,  
+    var cost: Double  
+){  
+  
+    fun crash(damage: Double){  
+        amountOfCrashes++  
+        cost -= damage  
+    }  
+  
+    fun sellTo(newOwner: String){  
+        ownerName = newOwner  
+    }  
+  
+    companion object{  
+        fun expensiveCar(ownerName: String) = Car(  
+            4,  
+            ownerName,  
+            0,  
+            1000000.0  
+        )  
+    }  
+  
+}  
+  
+fun main() {  
+    val carFields = Car::class.members.filterIsInstance<KProperty1<Car, *>>()  
+    val carMemberFunctions = Car::class.declaredMemberFunctions  
+    val carStaticFunctions = Car::class.companionObject?.declaredMemberFunctions  
+  
+    println("Class car has ${carFields.size} fields and ${carMemberFunctions.size} member functions")  
+    println("Fields are")  
+    carFields.forEach {  
+        println("    ${it.visibility} ${it.name}: ${it.returnType}")  
+    }  
+  
+    println("Member functions are")  
+    carMemberFunctions.forEach {  
+        val parameter_string = it.parameters.joinToString { param -> "${param.name}: ${param.type}" }  
+        println("    ${it.returnType} ${it.name} (${parameter_string})")  
+    }  
+  
+    println("Static functions are")  
+    carStaticFunctions?.forEach {  
+        val parameter_string = it.parameters.joinToString { param -> "${param.name}: ${param.type}" }  
+        println("    ${it.returnType} ${it.name} (${parameter_string})")  
+    }  
+}
+```
+Above you can see the simplest example of the reflection in Kotlin.
```

Comment:
Move this before example, because post started to talk about c++ and the first example is missleading